include/beman/scope/scope.hpp

```diff
@@ -113,7 +133,276 @@ make_unique_resource_checked(R&& r, const S& invalid, D&& d) noexcept(noexcept(
 //
 //        template <typename R, typename D>
 //        unique_resource(R, D) -> unique_resource<R, D>;
+//
+//==================================================================================================
+
+// Template argument `ScopeExitFunc` shall be
+//   - a function object type([function.objects]),
+//   - lvalue reference to function,
+//   - or lvalue reference to function object type.
+//
+//   If `ScopeExitFunc` is an object type, it shall meet the requirements of Cpp17Destructible(Table 30).
+//   Given an lvalue g of type remove_reference_t<EF>, the expression g() shall be well- formed.
+
+//==================================================================================================
+
+// --- Concepts ---
+template <typename F, typename R, typename... Args>
+concept invocable_return = std::invocable<F, Args...> && std::convertible_to<std::invoke_result_t<F, Args...>, R>;
+
+template <typename F>
+concept scope_exit_function =
+    invocable_return<F, void> && (std::is_nothrow_move_constructible_v<F> || std::is_copy_constructible_v<F>);
+
+template <typename T>
+concept scope_function_invoke_check = invocable_return<T, bool>;
+
+template <typename T>
+concept HasRelease = requires(T t) {
+    { t.release() } -> std::same_as<void>;
+};
+
+template <typename T>
+concept HasStaticRelease = requires {
+    { T::release() } -> std::same_as<void>;
+};
+
+// --- Enum ---
+
+enum class exception_during_constuction_behaviour : std::uint8_t {
+    dont_invoke_exit_func,
+    invoke_exit_func,
+};
+
+//==================================================================================================
+
+//  --- `scope_guard` - primary template ---
+
+template <scope_exit_function ScopeExitFunc,
+          typename InvokeChecker = void,
+          exception_during_constuction_behaviour ConstructionExceptionBehavior =
+              exception_during_constuction_behaviour::invoke_exit_func>
+class [[nodiscard]] scope_guard;
+
+//==================================================================================================
+
+// --- General definition ---
+
+template <scope_exit_function                    ScopeExitFunc,
+          scope_function_invoke_check            InvokeChecker,
+          exception_during_constuction_behaviour ConstructionExceptionBehavior>
+class [[nodiscard]] scope_guard<ScopeExitFunc, InvokeChecker, ConstructionExceptionBehavior> {
+  public:
+    // The constructor parameter `exit_func` in the following constructors shall be :
```

Comment:
```suggestion
    //* The constructor parameter `exit_func` in the following constructors shall be :
```
Let's make these comments doxygen viable -- during c++Now we started working on documentation tool chain that will allow doxygen comments to be generated into reference docs.